src/lib.rs

```diff
@@ -2439,6 +2456,38 @@ impl Build {
         Ok(())
     }
 
+    fn add_inherited_rustflags(&self, cmd: &mut Tool, target: &Target) -> Result<(), Error> {
+        let env_os = match self.getenv("CARGO_ENCODED_RUSTFLAGS") {
+            Some(env) => env,
+            // No encoded RUSTFLAGS -> nothing to do
+            None => return Ok(()),
+        };
+
+        let mut cc_flags: Vec<OsString> = env_os
+            .to_string_lossy()
+            .split("\u{1f}")
+            // Strip prefixes from rustc flags
+            .flat_map(|flag| {
+                if flag == "-Z" || flag == "-C" {
+                    None
+                } else if flag.starts_with("-Z") || flag.starts_with("-C") {
+                    Some(&flag[2..])
+                } else {
+                    Some(flag)
+                }
+            })
+            .flat_map(|flag| rustc_to_cc_flag(flag, cmd.family))
+            // Filter out flags not supported by the currently used CC
+            .filter(|flag| {
+                self.is_flag_supported_inner(flag, &cmd.path, &target)
+                    .unwrap_or(false)
+            })
+            .collect();
+
+        cmd.args.append(&mut cc_flags);
```

Comment:
I believe that using rust destruction will avoid the heap allocation and the lifetime error: 

```suggestion
        let Tool { family, path, args } = cmd;

        args.extend(env_os
            .to_string_lossy()
            .split("\u{1f}")
            // Strip prefixes from rustc flags
            .flat_map(|flag| {
                if flag == "-Z" || flag == "-C" {
                    None
                } else if flag.starts_with("-Z") || flag.starts_with("-C") {
                    Some(&flag[2..])
                } else {
                    Some(flag)
                }
            })
            .flat_map(|flag| rustc_to_cc_flag(flag, family))
            // Filter out flags not supported by the currently used CC
            .filter(|flag| {
                self.is_flag_supported_inner(flag, path, &target)
                    .unwrap_or(false)
            })
        );
```